walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -108,4 +109,83 @@
     Collections.sort(timeConflicts, TimeRange.ORDER_BY_START);
     return timeConflicts;
   }
+
+  /**
+   * Find the optimal time ranges for optional attendees.
+   *
+   * @param events
+   * @param request
+   * @return Returns the optimal time ranges for optional attendees. Can be empty.
+   */
+  private Collection<TimeRange> findOptimalOptionalAttendee(
+      Collection<Event> events, MeetingRequest request) {
+    // This works by checking each combination of optional attendees in increasing quantity.
+    // Starting with groups of one attendee, then two, etc. Until less than two combinations
+    // had non-empty ranges (see optimization notes below) or all combinations have been checked.
+    // 
+    // For large numbers of optional attendees, this can quickly become inefficient.
+    // There is room here for a dynamic programming approach following the logic that a combination
+    // of three elements could be more efficiently computed by using a previously-computed
+    // combination of two elements. 
+
+    ArrayList<String> optionalAttendees = 
+        new ArrayList<String>(request.getOptionalAttendees());
+    Collection<String> attendees = request.getAttendees();
+
+    int successfulCombinations = 0;
+    Collection<TimeRange> bestTimeRange = Arrays.asList();
+    int longestDuration = 0;
+    for (int i = 1; i <= optionalAttendees.size(); i++) {
+      Iterator<int[]> attendeeCombinations = 
+          CombinatoricsUtils.combinationsIterator(optionalAttendees.size(), i);
+      
+      // Find the longest duration of time ranges for each combination of i attendees.
+      while (attendeeCombinations.hasNext()) {
+        int[] combination = attendeeCombinations.next();
```

Comment:
Are you thinking of a for loop like [this](https://www.geeksforgeeks.org/for-each-loop-in-java/)? Because I was hoping I could do that but it turns out that it can only be done on something that returns an iterator, and not an iterator itself.
Or were you thinking of a for loop that's something like this:
`for (int[] combination = attendeeCombinations.next(); attendeeCombinations.hasNext(); combination = attendeeCombinations.next())`

Reply:
I was thinking of the first way, but since that isn't possible I like the solution you posted: `for (int[] combination = attendeeCombinations.next(); attendeeCombinations.hasNext(); combination = attendeeCombinations.next())`